### PR TITLE
Adding decimal literals

### LINKED ISSUE #90: implement decimal numeric literal

### PR DESCRIPTION
**Pull request by aupiff**
fixes #90 

READY TO MERGE.

NOTE:
- An ordinary decimal literal is assumed to to be of type `uint256`. To write a `uint8` literal one must add the suffix `u8`, as in `5u8`.

### CODE REVIEW

**Comment by aupiff** [comment]
I am working on automated gethtests in another [branch](https://github.com/aupiff/bamboo/commit/22d56d7c866bbbea9ab83ea1ee50b9907a161181).

I'll test various functions which use decimal literals on a local blockchain and then include those tests and the script to run them in separate PR. The `feature-automated-geth-testing` branch is in the very early stages, but it should be ready by the end of the week. I'll be confident about the correctness of this decimal literals PR by Thursday.

**Comment by pirapira** [comment]
Did you have a look at `src/exec/endToEnd.ml`?  It uses `eth` to test the compiled bytecode.

**Comment by pirapira** [comment]
The changes in this PR look good so far.

**Comment by aupiff** [comment]
@pirapira `src/exec/endToEnd.ml` looks good, but I have issues running `make endToEnd`.
```
ocamlbuild -use-ocamlfind -Is src/basics,src/ast,src/parse,src/lib,src/codegen -package batteries -package cryptokit -package rope -package rpclib -package rpclib.unix -package unix -package rpclib.json -package ppx_deriving -package ppx_deriving_rpc -package hex -use-menhir src/exec/endToEnd.native
Finished, 0 targets (0 cached) in 00:00:00.
+ /Users/aupiff/.opam/system/bin/ocamlfind ocamldep -package batteries -package cryptokit -package hex -package rope -package hex -package ppx_deriving_rpc -package ppx_deriving -package rpclib.json -package unix -package rpclib.unix -package rpclib -package rope -package cryptokit -package batteries -modules src/exec/endToEnd.ml > src/exec/endToEnd.ml.depends
ocamlfind: Package `ppx_deriving_rpc' not found                       
Command exited with code 2.        
Compilation unsuccessful after building 1 target (0 cached) in 00:00:00.
make: *** [endToEnd] Error 10
```

I'll write my own `geth` tests in bash for now and report on their results in a comment. Afterwards, I'll figure out how to compile add automated tests to `endToEnd.ml`

**Comment by aupiff** [comment]
Did some manually testing with `geth`, everything seems to work as expected. The `contract` below refers to `A()` from `025_declit_numeric.bbo`

```
> contract.multiply5.call(2)       
10               
> contract.multiply7.call(3)       
21               
> contract.multiply7.call(30000)   
210000           
> contract.s.call()                
true             
> contract.ggg.call()              
7.771213812919199999999e+22        
> contract.g.call(324234)          
10               
> contract.f.call(1000)            
false            
> contract.f.call(4)               
true             
> contract.f.call(5)               
false            
> contract.gg.call(23)             
12138138405817283711        
```

**Comment by pirapira** [comment]
The compilation error about `endToEnd` can be solved by this line: https://github.com/pirapira/bamboo/blob/master/.travis.yml#L40

Oh, now I notice (WIP) is gone.  I'll have a look.

**Comment by aupiff** [comment]
@pirapira just fixed some silly bugs and got `endToEnd` to compile, thanks for letting me know what deps I was missing.

**Comment by pirapira** [comment]
Documentation is missing about `endToEnd`.

**Comment by pirapira** [comment]
I created an issue #210 about the documentation.

File: src/ast/syntax.ml [hunk 1]
```diff
@@ -191,9 +193,11 @@ let string_of_exp_inner e =
   | IdentifierExp str -> "ident "^str
   | FunctionCallExp _ -> "call"
   | NowExp -> "now"
-  | FalseExp -> "false"
   | SenderExp -> "sender"
   | TrueExp -> "true"
+  | FalseExp -> "false"
```
Comment (pirapira) [inline src/ast/syntax.ml]: :+1:

File: src/parse/lexer.mll [hunk 2]
```diff
@@ -70,5 +70,10 @@ rule read =
   | "event" { EVENT }
   | "log" { LOG }
   | "indexed" { INDEXED }
+  | digit+ as i { DECLIT256 (Big_int.big_int_of_string i) }
+  (* uint has at most three digits *)
```
Comment (pirapira) [inline src/parse/lexer.mll]: `(* uint8 has at most three digits *)`

File: src/parse/negative_examples/uint8_too_big.bbo [hunk 3]
```diff
@@ -0,0 +1,7 @@
+contract A ()
+{
+    case(bool f(uint8 a))
+    {
+        return a < 300u8 then become A();
+    }
+}
```
Comment (pirapira) [inline src/parse/negative_examples/uint8_too_big.bbo]: Will you add a similar `uint256_too_big.bbo`?

**Comment by pirapira** [comment]
I'll deal with my own comments.

**Pull request merged**